### PR TITLE
[Observability] Added object refs Task is dependent on to `TaskInfoEntry`

### PR DESCRIPTION
**Pull request by alexeykudinkin**
## Why are these changes needed?

Currently when listing current tasks using `ray get task <task-id>` we're not reflecting the args the task might be dependent on and therefore waiting to become available.

That substantially complicates troubleshooting when the task is blocked on an argument passed in as an object-ref.

This change adds arguments passed in by reference to the `TaskInfoEntry` to subsequently make it available to our StateHead APIs.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -637,6 +637,7 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
         [this] {
           RAY_LOG(INFO) << "Event stats:\n\n"
                         << io_service_.stats().StatsString() << "\n\n"
+                        << task_execution_service_.stats().StatsString() << "\n\n"
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:640]: Haha, I have another PR that also adds this.

**Review by jjyao — COMMENTED** [review]
Thanks for doing that! Could you add some tests?

File: src/ray/protobuf/common.proto [hunk 2]
```diff
@@ -589,6 +589,13 @@ message TaskInfoEntry {
   // If the task/actor is created within a placement group,
   // this value is configured.
   optional bytes placement_group_id = 26;
+  // Tasks arguments passed in as (object) references.
+  //
+  // NOTE: This list only contains `ObjectReference`s passed in as arguments
+  //       this task is dependent on and does NOT contain
+  //        - Args passed by value (inlined)
+  //        - ObjectRefs of the args passed by value
+  repeated ObjectReference dependent_args_refs = 27;
```
Comment (jjyao) [inline src/ray/protobuf/common.proto:598]: I think we can only record the ObjectID instead of the entire ObjectRef?

Comment (alexeykudinkin) [reply src/ray/protobuf/common.proto:598]: Surely, but why not entire `ObjectRef`?

Comment (jjyao) [reply src/ray/protobuf/common.proto:598]: I think we only care about ObjectID so trying to reduce memory usage: we need to store lots of events in GCS also each task can potentially have unlimited number of args.

**Comment by jjyao** [comment]
There is test failure.

File: src/ray/protobuf/common.proto [hunk 3]
```diff
@@ -589,6 +589,13 @@ message TaskInfoEntry {
   // If the task/actor is created within a placement group,
   // this value is configured.
   optional bytes placement_group_id = 26;
+  // Tasks arguments passed in as (object) references.
+  //
+  // NOTE: This list only contains `ObjectReference`s passed in as arguments
+  //       this task is dependent on and does NOT contain
+  //        - Args passed by value (inlined)
+  //        - ObjectRefs of the args passed by value
```
Comment (jjyao) [inline src/ray/protobuf/common.proto:597]: What does this mean?

File: python/ray/util/state/common.py [hunk 4]
```diff
@@ -760,6 +760,9 @@ class TaskState(StateSchema):
     error_message: Optional[str] = state_column(detail=True, filterable=False)
     # Is task paused by the debugger
     is_debugger_paused: Optional[bool] = state_column(detail=True, filterable=True)
+    # List of objects (passed in by ref as arguments) this task
+    # is dependent
```
Comment (jjyao) [inline python/ray/util/state/common.py:764]:
```suggestion
    # is dependent on
```

File: python/ray/tests/test_state_api.py [hunk 5]
```diff
@@ -1048,31 +1051,42 @@ async def test_api_manager_list_tasks_events(state_api_manager):
...
+        "worker_id": None,
+        "task_log_info": None,
+        "profiling_data": {},
+        "actor_id": None,
+        "is_debugger_paused": None,
+        "start_time_ms": 2000.0,
+        "task_id": "31323334",
+        "parent_task_id": "",
+        "args_object_ids": [
+            arg_ref.hex(),
```
Comment (jjyao) [inline python/ray/tests/test_state_api.py:1087]: Don't we store the binary format of the object id instead of hex?

**Review by jjyao — COMMENTED** [review]
LGTM

File: src/ray/protobuf/common.proto (hunk 3 above)
Comment (alexeykudinkin) [reply src/ray/protobuf/common.proto:597]: Inlined ojbects

File: python/ray/tests/test_state_api.py (hunk 5 above)
Comment (alexeykudinkin) [reply python/ray/tests/test_state_api.py:1087]: We decode in the API

**Review by jjyao — APPROVED** [review]

**Pull request merged**